File: agents/ten_packages/extension/minimax_tts/pcm.go
/**
 *
 * Agora Real Time Engagement
 * Created by XinHui Li in 2024.
 * Copyright (c) 2024 Agora IO. All rights reserved.
 *
 */
// An extension written by Go for TTS
package extension

import (
	"fmt"
	"log/slog"

	"ten_framework/ten"
)

type pcm struct {
	config *pcmConfig
}

type pcmConfig struct {
	BytesPerSample    int32
	Channel           int32
	ChannelLayout     uint64
	Name              string
	SampleRate        int32
	SamplesPerChannel int32
	Timestamp         int64
}

func defaultPcmConfig() *pcmConfig {
	return &pcmConfig{
		BytesPerSample:    2,
		Channel:           1,
		ChannelLayout:     1,
		Name:              "pcm_frame",
		SampleRate:        32000,
		SamplesPerChannel: 32000 / 100,
		Timestamp:         0,
	}
}

func newPcm(config *pcmConfig) *pcm {
	return &pcm{
		config: config,
	}
}

func (p *pcm) getPcmFrame(buf []byte) (pcmFrame ten.AudioFrame, err error) {
	pcmFrame, err = ten.NewAudioFrame(p.config.Name)
	if err != nil {
		slog.Error(fmt.Sprintf("NewAudioFrame failed, err: %v", err), logTag)
		return
	}

	// set pcm frame
	pcmFrame.SetBytesPerSample(p.config.BytesPerSample)
	pcmFrame.SetSampleRate(p.config.SampleRate)
	pcmFrame.SetChannelLayout(p.config.ChannelLayout)
	pcmFrame.SetNumberOfChannels(p.config.Channel)
	pcmFrame.SetTimestamp(p.config.Timestamp)
	pcmFrame.SetDataFmt(ten.AudioFrameDataFmtInterleave)
	pcmFrame.SetSamplesPerChannel(p.config.SamplesPerChannel)
	pcmFrame.AllocBuf(p.getPcmFrameSize())

	borrowedBuf, err := pcmFrame.LockBuf()
	if err != nil {
		slog.Error(fmt.Sprintf("LockBuf failed, err: %v", err), logTag)
		return
	}

	// copy data
	copy(borrowedBuf, buf)

	pcmFrame.UnlockBuf(&borrowedBuf)
	return
}

func (p *pcm) getPcmFrameSize() int {
	return int(p.config.SamplesPerChannel * p.config.Channel * p.config.BytesPerSample)
}

func (p *pcm) newBuf() []byte {
	return make([]byte, p.getPcmFrameSize())
}

func (p *pcm) send(tenEnv ten.TenEnv, buf []byte) (err error) {
	pcmFrame, err := p.getPcmFrame(buf)
	if err != nil {
		slog.Error(fmt.Sprintf("getPcmFrame failed, err: %v", err), logTag)
		return
	}

	// send pcm
	if err = tenEnv.SendAudioFrame(pcmFrame); err != nil {
		slog.Error(fmt.Sprintf("SendAudioFrame failed, err: %v", err), logTag)
		return
	}

	return
}

File: agents/ten_packages/extension/elevenlabs_tts_python/pcm.py
#
#
# Agora Real Time Engagement
# Created by XinHui Li in 2024-07.
# Copyright (c) 2024 Agora IO. All rights reserved.
#
#
from typing import Iterator
from ten import AudioFrame, TenEnv, AudioFrameDataFmt


class Pcm:
    def __init__(self, config) -> None:
        self.config = config

    def get_pcm_frame(self, buf: memoryview) -> AudioFrame:
        frame = AudioFrame.create(self.config.name)
        frame.set_bytes_per_sample(self.config.bytes_per_sample)
        frame.set_sample_rate(self.config.sample_rate)
        frame.set_number_of_channels(self.config.num_channels)
        frame.set_timestamp(self.config.timestamp)
        frame.set_data_fmt(AudioFrameDataFmt.INTERLEAVE)
        frame.set_samples_per_channel(self.config.samples_per_channel // self.config.channel)

        frame.alloc_buf(self.get_pcm_frame_size())
        frame_buf = frame.lock_buf()
        # copy data
        frame_buf[:] = buf
        frame.unlock_buf(frame_buf)

        return frame

    def get_pcm_frame_size(self) -> int:
        return (self.config.samples_per_channel * self.config.channel * self.config.bytes_per_sample)

    def new_buf(self) -> bytearray:
        return bytearray(self.get_pcm_frame_size())

    def read_pcm_stream(self, stream: Iterator[bytes], chunk_size: int) -> Iterator[bytes]:
        chunk = b""
        for data in stream:
            chunk += data
            while len(chunk) >= chunk_size:
                yield chunk[:chunk_size]
                chunk = chunk[chunk_size:]

        if chunk:
            yield chunk


class PcmConfig:
    def __init__(self) -> None:
        self.bytes_per_sample = 2
        self.channel = 1
        self.name = "pcm_frame"
        self.sample_rate = 16000
        self.samples_per_channel = 16000 // 100
        self.timestamp = 0